improve ssl locking by using add_lock_callback and rw locks

robguima opened the pull request:
@deweerdt is in the loop and has reviewed this

This PR improves openssl and libressl locking in two ways:
1. Using RW locks (pthread) for one of the lock types (similar to what openssl does internally)
2. Adding a lock_add callback which avoids `lock-add-unlock`. 

We've been running similar changes in different applications with success - especially the add_lock callback, but this would need to be tested under load.

I originally intended to implement dynamic locks as per comment in ssl.c. However, we've found that dynlocks are only used by the CHIL engine in openssl, which makes them pretty useless. * we can still add them if desirable - also not sure whether perhaps libre makes better use of them.

BTW, this seems to work with the bundled libressl as well. Wonder if a test for specific openssl version must be added though.

kazuho commented:
@robguima Thank you for the PR.

I am perfectly fine with merging the add_lock callback part; thank you for noticing that it is missing!

For the rwlock part, I have following questions / suggestions. Could you please address them?

* Would you mind elaborating why the use of rwlock is limited to `CRYPTO_LOCK_EX_DATA`? I'm curious why the approach is not used for all the locks.
* I'd suggest calling `pthread_rwlockaddr_setkind_np` on Linux to set rwlock to writer-preferred to avoid starvation, considering the fact that the default of the platform is reader-preferred
* If you think we should be afraid of compatibility, it might be a good idea to obtain a write-lock _unless_ when `CRYPTO_READ` is set (and obtain read-lock only when `CRYPTO_READ` is set).

> I originally intended to implement dynamic locks as per comment in ssl.c. However, we've found that dynlocks are only used by the CHIL engine in openssl, which makes them pretty useless.

How about writing it down as a source code comment? Actually I knew the fact (and that's why it has been left as a TODO). Now I'm feeling guilty knowing that you have spent some time on it.

robguima commented:
@kazuho you are welcome and great points!

I will deep deeper on the rw lock bit; could also add a comment as well (of why that particular lock).

very interesting point on pthread_rwlockaddr_setkind_np - will add that. The original implementation this was based on used trylocks (followed by a reg lock upon failure) which we thought would not buy us anything in h2o because the locks would be heavily contended.

You mean changing the assert to an if statement? Sure I can do that. Was actually more concerned about the add_lock to be honest, but a quick gland showed libre to be be very similar in the locking callbacks.

BTW, this should be good until openssl 1.1 is finally out - the locking has been completely revamped and they use a lot more thread-local storage esp in the error stack.

will add the comment about dynlocks thanks!

robguima commented:
Also, could also split the PRs, so we can add the add_lock callback right away, which seems to be the low hanging fruit anyway.

kazuho commented:
@robguima 
> I will deep deeper on the rw lock bit; could also add a comment as well (of why that particular lock).

Thank you! Please note that I am not requesting a detailed report for the matter. I just wanted to know why you handled `CRYPTO_LOCK_EX_DATA` as the special case.

My understanding is that rwlock causes less contentions but uses more CPU cycles when no contention is observed; so I thought your decision might be based on the observations of how the locks of OpenSSL are used.

> (compatibility)You mean changing the assert to an if statement? Sure I can do that.

Yes. The reason I thought of it is because to me [the man page](https://www.openssl.org/docs/man1.0.2/crypto/threads.html) seemed to be vague in if either of `CRYPTO_READ` or `CRYPTO_WRITE` must be set, or if both can be left unset.

> Was actually more concerned about the add_lock to be honest, but a quick glance showed libre to be be very similar in the locking callbacks.

To me the purpose and the documentation seemed clear enough that your code would work fine.

> Could also split the PRs, so we can add the add_lock callback right away, which seems to be the low hanging fruit anyway.

Sounds sensible to me.

robguima commented:
not clear as to why that particular lock type - the reason predates me.

So, I've found that only certainly CRYPTO_LOCK types are actually used in CRYPTO_r_lock() internal calls, including the EX_DATA which is also locked and unlocked on every session. So, I followed them and added them to the is_rw_lock() function in the last commit. I've also tried always using rw locks and both openssl and libre always set the mode correctly so had no problems with that.

changed the logic a bit to default to the write lock to be a bit more conservative on removing the assert.

It will take a bit of testing to find out whether these help at all. If they end up having little contention, doubt that considering they are indeed used in TLS termination, we could be more conservative in the rw use and/or try using the trylock versions. Personally I suspect the additional cost of the rw locks vs mutex locks (about 40% in my quick test) is going to be noise considering the costs of the locking contention (even if minimal).

kazuho commented:
@robguima 
> So, I've found that only certainly CRYPTO_LOCK types are actually used in CRYPTO_r_lock() internal calls, including the EX_DATA which is also locked and unlocked on every session. Added the locks that are used by the CRYPTO_r_lock() in the is_rw_lock() function (the last commit).

It might be better to have a flag in each lock that indicates whether it is a mutex lock or a rw lock. Using a switch statement in is_rw_lock() function makes me afraid of branch prediction failures.

> If they end up having little contention, doubt that considering they are indeed used in TLS termination, could be more conservative in the rw use and/or try using the trylock versions. Personally I suspect the additional cost of the rw locks vs mutex locks (about 40% in my quick test) is going to be noise considering the costs of the TLS and locking contention.

Agreed (expect for the trylock part; I am not sure if using trylock causes any difference, my understanding is that pthread_mutex_lock first tries to lock in the userspace and only when that fails then issues a syscall).

Anyways I am looking forward to see what you find.

robguima commented:
Interesting - yeah will do that sure.

first nothing breaks by using the rw locks: can run RW locks exclusively or the selected few (as the PR).

I was able to run some preliminary tests and have not seen any obvious advantage in using the rw locks at all - actually seeing a surprising and obvious performance drop. I don't yet trust my test results because I had too many other variables that could affect it - so I will clean it up and match the builds (though the bundled ssl was the same - not sure that's a good thing).

Thought my first impression wasn't a positive one.

Will run some ideas by @deweerdt tomorrow.

robguima commented:
Updated by previous post because I found a build problem, so the large difference was due to using different bundled openssl libs.

On your previous point about branch prediction failures, I would be surprised if it did that around locks. No prob doing the flag version if we go for this though.

Was able run some tests:
The test consisted of driving load via loopback to different h2o builds. h2o was terminating TLS and serving the same file. So, the load is always the same. And then lowering the number of available cores via taskset - to increase locking contention:
taskset -c 0,2,4,6
taskset -c 2,4,6
taskset -c 4,6

h2o is configured to run with 16 threads.

This PR's implementation, without writer preference seemed to outperform the others. Writer preference seemed to slow things a bit until the contention was very high (4,6) in which case the improvement was minimal (by the writer preference).

Considering the variance, these results need to be taken with a grain of salt. I was somewhat confident by this PR improvement because of a 6% improvement over the other implementations. Everything else, the other takes, well it was just too close to call and it's more of a feeling over multiple results and tests.

There were some reassuring consistencies though: for example the ex_data lock (exclusive) binary improved over stable, but less so than this pr with more rw locks.

used 5 diff binaries:
stable (master)
rw_locks (this pr)
rw_locks_wr (^^ plus writer preference)
all_rw_locks (no switch with writer preference)
all_rw_locks_rd: (^^ but reader preference)
ex_data: (only ex_data rw lock)

I have an idea of writing an ssl stresser with bio_pairs to make this kind of tests easier; just need to get to it. The only drawback of such tool is that it'd have to do both client and server at the same time (in memory).

robguima commented:
As per @deweerdt adding the perf top data here. The (perf top) data suggests an improvement, in using RW locks per this PR, as the locking calls appear lower (the read lock even disappears at times). 

My personal take is that this is not worth the trouble. The variation in my test results is greater than the perceived improvement, so it's fair to say I have been unable to see any improvement in my testing.

Had to lower the key size to 1kbit otherwise the locking becomes noise. The one test I have not done is resumption - don't have the right tools for it as h2load doesn't support it.

regular implementation with muxes (from the top):
```
 16.49%  libcrypto.so.1.0.0  [.] __rsaz_512_reduce
   8.74%  libcrypto.so.1.0.0  [.] __ecp_nistz256_mul_montq
   8.62%  libcrypto.so.1.0.0  [.] rsaz_512_sqr
   7.00%  libcrypto.so.1.0.0  [.] __ecp_nistz256_sqr_montq
   5.38%  libcrypto.so.1.0.0  [.] rsaz_512_mul_gather4
   3.88%  libcrypto.so.1.0.0  [.] sha256_block_data_order_avx
   2.97%  libcrypto.so.1.0.0  [.] getrn
   2.57%  libc-2.15.so        [.] malloc
   2.30%  libc-2.15.so        [.] _int_free
   2.26%  libc-2.15.so        [.] _int_malloc
   1.53%  libcrypto.so.1.0.0  [.] bn_sqr8x_internal
   1.49%  libc-2.15.so        [.] __memcpy_ssse3_back
   1.48%  libcrypto.so.1.0.0  [.] __rsaz_512_subtract
   1.16%  libcrypto.so.1.0.0  [.] BN_div
   0.95%  libcrypto.so.1.0.0  [.] OPENSSL_cleanse
   0.87%  libc-2.15.so        [.] __memset_sse2
   0.86%  libcrypto.so.1.0.0  [.] ecp_nistz256_point_double
   0.78%  libcrypto.so.1.0.0  [.] __ecp_nistz256_sub_fromq
   0.76%  libcrypto.so.1.0.0  [.] __ecp_nistz256_mul_by_2q
   0.74%  libpthread-2.15.so  [.] pthread_mutex_lock      <---------
   0.71%  libcrypto.so.1.0.0  [.] sha1_block_data_order_avx
   0.67%  libcrypto.so.1.0.0  [.] EVP_MD_CTX_cleanup
   0.62%  libcrypto.so.1.0.0  [.] bn_sub_words
   0.56%  libc-2.15.so        [.] malloc_consolidate.part.3
   0.53%  libcrypto.so.1.0.0  [.] EVP_MD_CTX_init
   0.51%  libpthread-2.15.so  [.] pthread_mutex_unlock   <---------
   0.50%  libcrypto.so.1.0.0  [.] bn_mul_words
   0.43%  libcrypto.so.1.0.0  [.] __ecp_nistz256_add_toq
   0.42%  libcrypto.so.1.0.0  [.] __rsaz_512_mul
   0.39%  libc-2.15.so        [.] free
   0.39%  libcrypto.so.1.0.0  [.] CRYPTO_malloc
   0.38%  libcrypto.so.1.0.0  [.] bn_mul4x_mont
   0.37%  libcrypto.so.1.0.0  [.] BN_lshift
   0.36%  libcrypto.so.1.0.0  [.] ecp_nistz256_point_add
   0.36%  libc-2.15.so        [.] vfprintf
   0.31%  libssl.so.1.0.0     [.] ssl3_read_bytes
   0.30%  libssl.so.1.0.0     [.] ssl3_accept
   0.30%  libcrypto.so.1.0.0  [.] ecp_nistz256_select_w5
   0.30%  libssl.so.1.0.0     [.] tls1_enc
   0.26%  libcrypto.so.1.0.0  [.] EVP_MD_CTX_copy_ex
   0.26%  libcrypto.so.1.0.0  [.] ecp_nistz256_points_mul
   0.24%  libcrypto.so.1.0.0  [.] BN_copy
   0.23%  libcrypto.so.1.0.0  [.] __ecp_nistz256_subq
   0.23%  libssl.so.1.0.0     [.] do_ssl3_write
   0.22%  libcrypto.so.1.0.0  [.] ASN1_item_ex_i2d
   0.21%  libcrypto.so.1.0.0  [.] OBJ_bsearch_ex_
   0.21%  libcrypto.so.1.0.0  [.] sha512_block_data_order_avx
   0.20%  libcrypto.so.1.0.0  [.] ecp_nistz256_sqr_mont
   0.20%  libcrypto.so.1.0.0  [.] SHA256_Update
   0.18%  libcrypto.so.1.0.0  [.] gcm_ghash_clmul
   0.16%  libcrypto.so.1.0.0  [.] aesni_ctr32_encrypt_blocks
   0.16%  libcrypto.so.1.0.0  [.] HMAC_CTX_cleanup
```

this PR's implementation:
```
  15.38%  libcrypto.so.1.0.0  [.] __rsaz_512_reduce
   9.82%  libcrypto.so.1.0.0  [.] __ecp_nistz256_mul_montq
   8.24%  libcrypto.so.1.0.0  [.] rsaz_512_sqr
   7.92%  libcrypto.so.1.0.0  [.] __ecp_nistz256_sqr_montq
   4.88%  libcrypto.so.1.0.0  [.] rsaz_512_mul_gather4
   3.94%  libcrypto.so.1.0.0  [.] sha256_block_data_order_avx
   2.90%  libcrypto.so.1.0.0  [.] getrn
   2.33%  libc-2.15.so        [.] malloc
   2.30%  libc-2.15.so        [.] _int_malloc
   2.28%  libc-2.15.so        [.] _int_free
   1.46%  libcrypto.so.1.0.0  [.] __rsaz_512_subtract
   1.43%  libc-2.15.so        [.] __memcpy_ssse3_back
   1.42%  libcrypto.so.1.0.0  [.] bn_sqr8x_internal
   1.13%  libcrypto.so.1.0.0  [.] BN_div
   0.93%  libcrypto.so.1.0.0  [.] ecp_nistz256_point_double
   0.89%  libc-2.15.so        [.] __memset_sse2
   0.87%  libcrypto.so.1.0.0  [.] OPENSSL_cleanse
   0.86%  libcrypto.so.1.0.0  [.] __ecp_nistz256_sub_fromq
   0.86%  libcrypto.so.1.0.0  [.] __ecp_nistz256_mul_by_2q
   0.69%  libcrypto.so.1.0.0  [.] sha1_block_data_order_avx
   0.68%  libcrypto.so.1.0.0  [.] bn_sub_words
   0.67%  libcrypto.so.1.0.0  [.] EVP_MD_CTX_cleanup
   0.64%  libc-2.15.so        [.] malloc_consolidate.part.3
   0.48%  libcrypto.so.1.0.0  [.] EVP_MD_CTX_init
   0.46%  libcrypto.so.1.0.0  [.] bn_mul_words
   0.43%  libcrypto.so.1.0.0  [.] __ecp_nistz256_add_toq
   0.43%  libc-2.15.so        [.] vfprintf
   0.43%  libcrypto.so.1.0.0  [.] bn_mul4x_mont
   0.42%  libcrypto.so.1.0.0  [.] ecp_nistz256_point_add
   0.38%  libcrypto.so.1.0.0  [.] CRYPTO_malloc
   0.37%  libpthread-2.15.so  [.] pthread_rwlock_unlock     <---------
   0.37%  libcrypto.so.1.0.0  [.] ecp_nistz256_select_w5
   0.34%  libcrypto.so.1.0.0  [.] __rsaz_512_mul
   0.32%  libcrypto.so.1.0.0  [.] ecp_nistz256_points_mul
   0.31%  libssl.so.1.0.0     [.] tls1_enc
   0.31%  libssl.so.1.0.0     [.] ssl3_read_bytes
   0.31%  libc-2.15.so        [.] free
   0.30%  libcrypto.so.1.0.0  [.] EVP_MD_CTX_copy_ex
   0.28%  h2o.final           [.] lock_callback
   0.28%  libssl.so.1.0.0     [.] ssl3_accept
   0.25%  libcrypto.so.1.0.0  [.] BN_lshift
   0.25%  h2o.final           [.] add_lock_callback
   0.25%  libcrypto.so.1.0.0  [.] __ecp_nistz256_subq
   0.23%  libcrypto.so.1.0.0  [.] OBJ_bsearch_ex_
   0.22%  libcrypto.so.1.0.0  [.] BN_copy
   0.21%  libssl.so.1.0.0     [.] do_ssl3_write
   0.20%  libcrypto.so.1.0.0  [.] ASN1_item_ex_i2d
   0.20%  libcrypto.so.1.0.0  [.] aesni_ctr32_encrypt_blocks
   0.18%  libcrypto.so.1.0.0  [.] sha512_block_data_order_avx
   0.18%  libcrypto.so.1.0.0  [.] gcm_ghash_clmul
   0.18%  libcrypto.so.1.0.0  [.] ecp_nistz256_sqr_mont
   0.17%  libpthread-2.15.so  [.] pthread_rwlock_wrlock     <---------
   0.17%  libpthread-2.15.so  [.] pthread_rwlock_rdlock     <---------
```

kazuho commented:
@robguima Thank you for the updates.

It's unfortunate that using RW locks didn't prove improvements. But I think the work was a great experiment.

Let's leave this PR as is, so that people using other environments (e.g. ARM) can look into testing it. Also, we might need to reevaluate this PR in the future as the number of cores increases.

robguima commented:
@kazuho you are welcome

when/If I implement the openssl tester (might also be handy esp come 1.1) I will re-test and post results here.

thank you
